src/Versions.jl

```diff
@@ -8,34 +8,37 @@ export VersionBound, VersionRange, VersionSpec, semver_spec, isjoinable
 # VersionBound #
 ################
 struct VersionBound
-    t::NTuple{3,UInt32}
+    t::NTuple{4,UInt32}
     n::Int
     function VersionBound(tin::NTuple{n,Integer}) where n
-        n <= 3 || throw(ArgumentError("VersionBound: you can only specify major, minor and patch versions"))
-        n == 0 && return new((0,           0,      0), n)
-        n == 1 && return new((tin[1],      0,      0), n)
-        n == 2 && return new((tin[1], tin[2],      0), n)
-        n == 3 && return new((tin[1], tin[2], tin[3]), n)
+        n <= 4 || throw(ArgumentError("VersionBound: you can only specify major, minor, patch versions, and a build number"))
+        n == 0 && return new((0,           0,      0,      0), n)
+        n == 1 && return new((tin[1],      0,      0,      0), n)
+        n == 2 && return new((tin[1], tin[2],      0,      0), n)
+        n == 3 && return new((tin[1], tin[2], tin[3],      0), n)
+        n == 4 && return new((tin[1], tin[2], tin[3], tin[4]), n)
         error("invalid $n")
     end
 end
 VersionBound(t::Integer...) = VersionBound(t)
-VersionBound(v::VersionNumber) = VersionBound(v.major, v.minor, v.patch)
+VersionBound(v::VersionNumber) = VersionBound(v.major, v.minor, v.patch, v.build)
```

Comment:
Or, to get even fancier:

```suggestion
function VersionBound(v::VersionNumber)
    if isempty(v.build)
        return VersionBound(v.major, v.minor, v.patch)
    elseif length(v.build) > 1
            pkgerror("We only support a single build number; multiple build numbers are not supported")
    else
        if !(v.build isa Integer)
            pkgerror("We only support integer values for the build data field")
        end
        return VersionBound(v.major, v.minor, v.patch, only(v.build))
    end
end
```

Reply:
There is an argument that they should silently ignore it all, which is what Pkg currently does
```
julia> Pkg.Types.VersionBound(v"1.2.3+4")
Pkg.Versions.VersionBound((0x00000001, 0x00000002, 0x00000003), 3)

julia> Pkg.Types.VersionBound(v"1.2.3+4.5")
Pkg.Versions.VersionBound((0x00000001, 0x00000002, 0x00000003), 3)

julia> Pkg.Types.VersionBound(v"1.2.3+4.5.6")
Pkg.Versions.VersionBound((0x00000001, 0x00000002, 0x00000003), 3)

julia> Pkg.Types.VersionBound(v"1.2.3+foo")
Pkg.Versions.VersionBound((0x00000001, 0x00000002, 0x00000003), 3)
```

Reply:
Maybe print a `@debug`?